File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/recipes/persistence/repository/DatabaseRecipeRepository.java
package com.tarbadev.witchcraft.recipes.persistence.repository;

import com.tarbadev.witchcraft.recipes.domain.entity.Ingredient;
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe;
import com.tarbadev.witchcraft.recipes.domain.repository.RecipeRepository;
import com.tarbadev.witchcraft.recipes.persistence.entity.IngredientEntity;
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity;
import com.tarbadev.witchcraft.recipes.persistence.entity.StepEntity;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class DatabaseRecipeRepository implements RecipeRepository {
  private final RecipeEntityRepository recipeEntityRepository;

  public DatabaseRecipeRepository(RecipeEntityRepository recipeEntityRepository) {
    this.recipeEntityRepository = recipeEntityRepository;
  }

  @Override
  public Recipe saveRecipe(Recipe recipe) {
    return recipeEntityRepository.saveAndFlush(new RecipeEntity(recipe)).recipe();
  }

  @Override
  public Recipe updateRecipe(Recipe recipe) {
    RecipeEntity entity = recipeEntityRepository.findById(recipe.getId()).orElse(null);
    entity.setName(recipe.getName());
    entity.setOriginUrl(recipe.getOriginUrl());
    entity.setImgUrl(recipe.getImgUrl());
    entity.setIngredients(recipe.getIngredients().stream().map(IngredientEntity::new).collect(Collectors.toList()));
    entity.setSteps(recipe.getSteps().stream().map(StepEntity::new).collect(Collectors.toList()));
    return recipeEntityRepository.saveAndFlush(entity).recipe();
  }

  @Override
  public List<Recipe> findAll() {
    return recipeEntityRepository.findAllByOrderByName().stream()
        .map(RecipeEntity::recipe)
        .collect(Collectors.toList());
  }

  @Override
  public Recipe findById(Integer id) {
    return recipeEntityRepository.findById(id)
        .map(RecipeEntity::recipe)
        .map(recipe -> {
          recipe.getIngredients().sort(Comparator.comparing(Ingredient::getName));
          return recipe;
        })
        .orElse(null);
  }

  @Override
  public void delete(int id) {
    recipeEntityRepository.deleteById(id);
    recipeEntityRepository.flush();
  }

  @Override
  public void setFavorite(int id, Boolean favorite) {
    RecipeEntity recipeEntity = recipeEntityRepository.findById(id).get();
    recipeEntity.setFavorite(favorite);
    recipeEntityRepository.flush();
  }

  @Override
  public List<Recipe> findAllFavorite() {
    return recipeEntityRepository.findAllByFavorite(true).stream()
        .map(RecipeEntity::recipe)
        .collect(Collectors.toList());
  }

  @Override
  public List<Recipe> findLastAddedRecipes() {
    return recipeEntityRepository.findTop8ByOrderByIdDesc().stream()
        .map(RecipeEntity::recipe)
        .collect(Collectors.toList());
  }

  @Override
  public Boolean existsById(Integer id) {
    return recipeEntityRepository.existsById(id);
  }
}

File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/recipes/persistence/repository/DatabaseRecipeRepository.kt
package com.tarbadev.witchcraft.recipes.persistence.repository

import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.recipes.domain.repository.RecipeRepository
import com.tarbadev.witchcraft.recipes.persistence.entity.IngredientEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.StepEntity
import org.springframework.stereotype.Component

@Component
class DatabaseRecipeRepository(private val recipeEntityRepository: RecipeEntityRepository) : RecipeRepository {

    override fun saveRecipe(recipe: Recipe): Recipe {
        return recipeEntityRepository.saveAndFlush(RecipeEntity(recipe)).recipe()
    }

    override fun updateRecipe(recipe: Recipe): Recipe {
        val entity = recipeEntityRepository.findById(recipe.id).orElse(null).copy(
            name = recipe.name,
            originUrl = recipe.originUrl,
            imgUrl = recipe.imgUrl,
            ingredients = recipe.ingredients.map { IngredientEntity(it) },
            steps = recipe.steps.map { StepEntity(it) }.toSet()
        )
        return recipeEntityRepository.saveAndFlush(entity).recipe()
    }

    override fun findAll(): List<Recipe> {
        return recipeEntityRepository.findAllByOrderByName()
            .filter { !it.isArchived }
            .map { it.recipe() }
    }

    override fun findById(id: Int): Recipe? {
        return recipeEntityRepository.findById(id)
            .filter { !it.isArchived }
            .map { it.recipe() }
            .orElse(null)
    }

    override fun delete(id: Int) {
        recipeEntityRepository.findById(id).get().isArchived = true
        recipeEntityRepository.flush()
    }

    override fun setFavorite(id: Int, isFavorite: Boolean): Recipe {
        val recipeEntity = recipeEntityRepository.findById(id).get()
        recipeEntity.favorite = isFavorite

        recipeEntityRepository.flush()

        return recipeEntity.recipe()
    }

    override fun findAllFavorite(): List<Recipe> {
        return recipeEntityRepository.findAllByFavorite(true)
            .filter { !it.isArchived }
            .map { it.recipe() }
    }

    override fun findLastAddedRecipes(): List<Recipe> {
        return recipeEntityRepository.findTop8ByOrderByIdDesc()
            .filter { !it.isArchived }
            .map { it.recipe() }
    }

    override fun existsById(id: Int): Boolean {
        return recipeEntityRepository.existsById(id)
    }
}
